### ISSUE TITLE
No satellite when loading robot with xacro

### ISSUE THREAD

**Issue by linuxsocke** [PX4/PX4-SITL_gazebo-classic#660]
Hi, 

I am currently loading in my robot description  with xacro and my mavlink doesnt see any satellites. I know xacro is now deprecated (why?) but is there any chance that it can be used aswell in the future? 

Loading in the libgazebo gps plugin looks like this in my xacros:

![image](https://user-images.githubusercontent.com/49939709/99902291-ff156c80-2cbc-11eb-8a0f-4ece93e25084.png)
 
Looking into diagnotics I see:

    level: 0
    name: "convecdro_0/mavros: FCU connection"
    message: "connected"
    hardware_id: "udp://:14540@localhost:14580"
    values: 
      - 
        key: "Received packets:"
        value: "38390"
      - 
        key: "Dropped packets:"
        value: "0"
      - 
        key: "Buffer overruns:"
        value: "0"
      - 
        key: "Parse errors:"
        value: "0"
      - 
        key: "Rx sequence number:"
        value: "246"
      - 
        key: "Tx sequence number:"
        value: "0"
      - 
        key: "Rx total bytes:"
        value: "1238559"
      - 
        key: "Tx total bytes:"
        value: "268575"
      - 
        key: "Rx speed:"
        value: "2413.000000"
      - 
        key: "Tx speed:"
        value: "537.000000"
      - 
    level: 2
    name: "convecdro_0/mavros: GPS"
    message: "No satellites"
    hardware_id: "udp://:14540@localhost:14580"
    values: 
      - 
        key: "Satellites visible"
        value: "0"
      - 
        key: "Fix type"
        value: "0"
      - 
        key: "EPH (m)"
        value: "Unknown"
      - 
        key: "EPV (m)"
        value: "Unknown"
      - 
    level: 0
    name: "convecdro_0/mavros: Heartbeat"
    message: "Normal"
    hardware_id: "udp://:14540@localhost:14580"
    values: 
      - 
        key: "Heartbeats since startup"
        value: "495"
      - 
        key: "Frequency (Hz)"
        value: "1.000000"
      - 
        key: "Vehicle type"
        value: "Hexarotor"
      - 
        key: "Autopilot type"
        value: "PX4 Autopilot"
      - 
        key: "Mode"
        value: "MANUAL"
      - 
        key: "System status"
        value: "Standby"
      - 
    level: 2
    name: "convecdro_0/mavros: System"
    message: "Sensor health"
    hardware_id: "udp://:14540@localhost:14580"
    values: 
      - 
        key: "Sensor present"
        value: "0x1008002C"
      - 
        key: "Sensor enabled"
        value: "0x1021002C"
      - 
        key: "Sensor health"
        value: "0x000E002C"
      - 
        key: "3D magnetometer"
        value: "Ok"
      - 
        key: "absolute pressure"
        value: "Ok"
      - 
        key: "GPS"
        value: "Ok"
      - 
        key: "rc receiver"
        value: "Fail"
      - 
        key: "AHRS subsystem health"
        value: "Fail"

And here some of my output:

Gazebo multi-robot simulator, version 11.1.0
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Wrn] [GuiIface.cc:199] g/gui-plugin is really loading a SystemPlugin. To load a GUI plugin please use --gui-client-plugin 
Gazebo multi-robot simulator, version 11.1.0
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[ INFO] [1606044976.934388707]: Finished loading Gazebo ROS API Plugin.
[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 192.168.0.108
[ INFO] [1606044976.949512557]: waitForService: Service [/gazebo/set_physics_properties] has not been advertised, waiting...
[INFO] [1606044976.956348, 0.000000]: Loading model XML from ros parameter robot_description
[INFO] [1606044976.958767, 0.000000]: Waiting for service /gazebo/spawn_urdf_model
[ INFO] [1606044977.072696561]: Finished loading Gazebo ROS API Plugin.
[ INFO] [1606044977.073675664]: waitForService: Service [/gazebo_gui/set_physics_properties] has not been advertised, waiting...
[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 192.168.0.108
[ INFO] [1606044978.673960327]: waitForService: Service [/gazebo/set_physics_properties] is now available.
[ INFO] [1606044978.696094166, 0.004000000]: Physics dynamic reconfigure ready.
[Wrn] [GuiIface.cc:119] Qt: Session management error: None of the authentication protocols specified are supported
[INFO] [1606044978.774081, 0.000000]: Calling service /gazebo/spawn_urdf_model
[INFO] [1606044978.845541, 0.152000]: Calling service /gazebo/spawn_urdf_model
[INFO] [1606044978.895895, 0.164000]: Spawn status: SpawnModel: Successfully spawned entity
[Wrn] [msgs.cc:1841] Conversion of sensor type[gpu_ray] not supported.
[INFO] [1606044979.225870, 0.368000]: Spawn status: SpawnModel: Successfully spawned entity
[ INFO] [1606044979.428465973, 0.368000000]: Laser Plugin: Using the 'robotNamespace' param: 'convecdro_0'
[ INFO] [1606044979.428557426, 0.368000000]: Starting GazeboRosLaser Plugin (ns = convecdro_0)
[ INFO] [1606044979.441788332, 0.368000000]: GPU Laser Plugin (ns = convecdro_0) <tf_prefix_>, set to "convecdro_0"
[ INFO] [1606044979.442472124, 0.368000000]: LoadThread function completed
[Msg] Connecting to PX4 SITL using TCP
[Msg] Lockstep is enabled
[Msg] Speed factor set to: 1
[Wrn] [gazebo_mavlink_interface.cpp:94] No identifier on joint. Using 0 as default sensor ID
[Msg] Using MAVLink protocol v2.0
INFO  [simulator] Simulator connected on TCP port 4560.
[beacon_0/spawn_beacon_0-10] process has finished cleanly
log file: /home/linuxsocke/.ros/log/edc08ccc-2cb6-11eb-b2b3-aff02fa65868/beacon_0-spawn_beacon_0-10*.log
INFO  [commander] LED: open /dev/led0 failed (22)
[Wrn] [msgs.cc:1841] Conversion of sensor type[gpu_ray] not supported.
[Wrn] [msgs.cc:1841] Conversion of sensor type[gpu_ray] not supported.
INFO  [init] Mixer: etc/mixers/hexa_x.main.mix on /dev/pwm_output0
INFO  [ekf2] starting instance 0, IMU:0 (1310988), MAG:0 (197388)
INFO  [ekf2] starting instance 1, IMU:1 (1310996), MAG:0 (197388)
INFO  [ekf2] starting instance 2, IMU:2 (1311004), MAG:0 (197388)
INFO  [ekf2] starting instance 3, IMU:0 (1310988), MAG:1 (197644)
INFO  [ekf2] starting instance 4, IMU:1 (1310996), MAG:1 (197644)
INFO  [ekf2] starting instance 5, IMU:2 (1311004), MAG:1 (197644)
INFO  [mavlink] mode: Normal, data rate: 4000000 B/s on udp port 18570 remote port 14550
[convecdro_0/spawn_convecdro_0-5] process has finished cleanly
log file: /home/linuxsocke/.ros/log/edc08ccc-2cb6-11eb-b2b3-aff02fa65868/convecdro_0-spawn_convecdro_0-5*.log
INFO  [mavlink] mode: Onboard, data rate: 4000000 B/s on udp port 14580 remote port 14540
INFO  [mavlink] mode: Onboard, data rate: 4000 B/s on udp port 14280 remote port 14030
INFO  [mavlink] mode: Onboard, data rate: 4000 B/s on udp port 14558 remote port 14530
INFO  [logger] logger started (mode=all)
INFO  [logger] Start file log (type: full)
INFO  [logger] [logger] ./log/2020-11-22/11_36_20.ulg
INFO  [logger] Opened full log file: ./log/2020-11-22/11_36_20.ulg
INFO  [mavlink] MAVLink only on localhost (set param MAV_BROADCAST = 1 to enable network)
INFO  [px4] Startup script returned successfully
INFO  [mavlink] partner IP: 127.0.0.1
[ INFO] [1606044980.113169141, 0.860000000]: udp0: Remote address: 127.0.0.1:14580
[ INFO] [1606044981.008520006, 1.760000000]: CON: Got HEARTBEAT, connected. FCU: PX4 Autopilot
[Wrn] [Publisher.cc:136] Queue limit reached for topic /gazebo/convecdro_0/convecdro_0/gazebo/command/motor_speed, deleting message. This warning is printed only once.
[ INFO] [1606044982.018057997, 2.768000000]: WP: Using MISSION_ITEM_INT
[ INFO] [1606044982.018473878, 2.768000000]: VER: 1.1: Capabilities         0x000000000000e4ef
[ INFO] [1606044982.019039602, 2.768000000]: VER: 1.1: Flight software:     010b00c0 (5f0c572c61000000)
[ INFO] [1606044982.019078950, 2.768000000]: VER: 1.1: Middleware software: 010b00c0 (5f0c572c61000000)
[ INFO] [1606044982.019110929, 2.768000000]: VER: 1.1: OS software:         050400ff (904a602c74dc08a1)
[ INFO] [1606044982.019155718, 2.768000000]: VER: 1.1: Board hardware:      00000001
[ INFO] [1606044982.019189329, 2.768000000]: VER: 1.1: VID/PID:             0000:0000
[ INFO] [1606044982.019227195, 2.768000000]: VER: 1.1: UID:                 4954414c44494e4f
[ INFO] [1606044991.020732451, 11.760000000]: HP: requesting home position
[ INFO] [1606044996.019863469, 16.760000000]: WP: mission received
[ INFO] [1606045001.023309110, 21.760000000]: HP: requesting home position
[ INFO] [1606045011.031395473, 31.760000000]: HP: requesting home position
[ INFO] [1606045021.038888510, 41.760000000]: HP: requesting home position

If you want I can also show the launch files and how I am loading the mavlink interface plugin.

**System Information**
Ubuntu: 20.04
ROS: Neotic
PX4: 1.11.2

**Comment by Jaeyoung-Lim** [comment]
@linuxsocke Because it is unable to talk to the `gazebo_mavlink_interface`, which sends the gps values to PX4 throught the HIL_GPS message. 

You need to attach your gps plugin to a sensor so that the mavlink interface plugin can figure out that it exists in your model. You can check the code how this works. https://github.com/PX4/PX4-SITL_gazebo/blob/a6698f5e7b4faa68b1ea297f821b15020aa4545e/src/gazebo_mavlink_interface.cpp#L382

**Comment by linuxsocke** [comment]
@Jaeyoung-Lim I want to spawn a robot in ros-gazebo by providing a description using xacro. 

So when adding a gps it expects a gps_joint in the set of joints of the robot? I already tried to add a gps_link and a joint which connects the link to the base_link. I called the joint like ${namespace}/gps_joint or simply gps_joint. Or am I just completely getting something wrong?

**Comment by Jaeyoung-Lim** [comment]
@linuxsocke I don't think xacro will work at all with this repo, unless you convert the generated urdf file into a sdf file. 

Have you looked at the code I have linked above? You can see how the topic names it expects from the gps plugin

**Comment by linuxsocke** [comment]
Ah ok I didnt see. Its expecting only sdf format.

So a sdf template for jinja is the only way.

**Comment by Jaeyoung-Lim** [comment]
@linuxsocke You can also create a `urdf` file from xacro, then covert the urdf file to a sdf file through `gz sdf -p`.

However, I am not sure if this would be worth the effort

Also using jinja templates are not "necessary". There are a lot of models in this repo that use sdf files directly and not use the jinja templates

**Comment by linuxsocke** [comment]
I think so too. It would be another story if the conversion can be somehow done also in the launch process.

So with a jinja template my custom mav has satellite now.

What are the advantages of jinja over simply using sdf files? I know it can be used to replace some parameters. But how do I e.g. set namespaces for my links?

**Comment by Jaeyoung-Lim** [comment]
jinja templates are there to configure runtime configurations such as TCP ports, UART devices .etc and generate a sdf file from it.

For your use case you don't need to use jinja templates if you don't need it.

**Comment by crazy-robotics** [comment]
@linuxsocke  Have you solved it? I am getting `No valid position estimation` error because I don't have GPS,  following plugin is called in `libgazebo_gps_plugin.so` . Here is the [xacro file ](https://github.com/crazy-robotics/dronedoc/blob/master/dronedoc/models/iris_depth_camera/xacro_sitl/urdf/component_snippets.xacro#L127)

**Comment by crazy-robotics** [comment]
@linuxsocke I was facing the same error [ref1](https://discuss.px4.io/t/no-valid-position-estimate/44476) [ref2](https://discuss.px4.io/t/gps-plugin-not-working-with-xacro-file-libgazebo-gps-plugin-so/44453) I was using default gps plugin 
```
  <xacro:macro name="gps_plugin_macro" params="namespace gps_noise">
    <gazebo>
      <plugin name='gps_plugin' filename='libgazebo_gps_plugin.so'>
        <robotNamespace/>
        <gpsNoise>1</gpsNoise>
        <linkName>gps0</linkName>
        <gpsSubTopic>/gps</gpsSubTopic>
      </plugin>
    </gazebo>
  </xacro:macro>
```
I was getting position error 

![Image](https://github.com/user-attachments/assets/4fc122af-1406-4f64-b67f-65021ccca7a4)


then I changed to following 
```
  <xacro:macro name="gps_plugin_macro" params="namespace gps_noise">
    <gazebo>
    <sensor name='gps' type='gps'>
    <pose>0 0 0 0 0 0</pose>
    <update_rate>5.0</update_rate>
    <always_on>true</always_on>
    <visualize>false<visualize>
      <plugin name='gps_plugin' filename='libgazebo_gps_plugin.so'>
        <robotNamespace/>
        <gpsNoise>1</gpsNoise>
        <gpsXYRandomWalk>2.0</gpsXYRandomWalk>
        <gpsZRandomWalk>4.0</gpsZRandomWalk>
        <gpsXYNoiseDensity>0.0002</gpsXYNoiseDensity>
        <gpsZNoiseDensity>0.0004</gpsZNoiseDensity>
        <gpsVXYNoiseDensity>0.2</gpsVXYNoiseDensity>
        <gpsZNoiseDensity>0.4</gpsZNoiseDensity>
        <linkName>gps0</linkName>
        <gpsSubTopic>/gps</gpsSubTopic>
      </plugin>
    </gazebo>
  </xacro:macro>
``` 
the error was solved 
![Image](https://github.com/user-attachments/assets/a2f31663-1ec5-4581-b26a-82ad4d141bdb)
also GPS count 

![Image](https://github.com/user-attachments/assets/6be7537b-bfdc-4019-9677-2b0a02d16564)

**Comment by DronecodeBot** [comment]
This issue has been mentioned on **Discussion Forum for PX4, Pixhawk, QGroundControl, MAVSDK, MAVLink**. There might be relevant details there:

https://discuss.px4.io/t/gps-plugin-not-working-with-xacro-file-libgazebo-gps-plugin-so/44453/2
